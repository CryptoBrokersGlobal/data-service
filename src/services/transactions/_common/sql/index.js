const {
  compose,
  map,
  pick,
  filter,
  has,
  __,
  identity,
  reverse,
} = require('ramda');

const commonFiltersOrder = require('./filtersOrder');
const commonFilters = require('./filters');

const createSql = ({
  query,
  filters = commonFilters,
  filtersOrder = commonFiltersOrder,
  queryAfterFilters = identity,
}) => ({
  get: id =>
    compose(
      String,
      queryAfterFilters,
      filters.id(id)
    )(query),

  mget: ids =>
    compose(
      String,
      queryAfterFilters,
      filters.ids(ids)
    )(query),

  search: fValues => {
    const fValuesPicked = pick(filtersOrder, fValues);

    const appliedFs = compose(
      map(x => filters[x](fValuesPicked[x])),
      filter(has(__, fValuesPicked)),
      reverse
    )(filtersOrder);

    return compose(
      String,
      queryAfterFilters,
      ...appliedFs,
      q => q.clone()
    )(query);
  },
});

module.exports = createSql;
