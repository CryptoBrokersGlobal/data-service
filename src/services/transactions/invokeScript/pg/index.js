const Maybe = require('folktale/maybe');
const { head, propEq } = require('ramda');
const { toDbError } = require('../../../../errorHandling');

const { matchRequestsResults } = require('../../../../utils/db');

const sql = require('./sql');
const transformResult = require('./transformResult');

export default {
  get: pg => id =>
    pg
      .any(sql.get(id))
      .map(transformResult)
      .map(head)
      .map(Maybe.fromNullable)
      .mapRejected(
        toDbError({ request: 'transactions.invokeScript.get', params: id })
      ),

  mget: pg => ids =>
    pg
      .any(sql.mget(ids))
      .map(transformResult)
      .map(matchRequestsResults(propEq('id'), ids))
      .mapRejected(
        toDbError({ request: 'transactions.invokeScript.mget', params: ids })
      ),

  search: pg => filters =>
    pg
      .any(sql.search(filters))
      .map(transformResult)
      .mapRejected(
        toDbError({
          toDbError: 'transactions.invokeScript.search',
          params: filters,
        })
      ),
};
