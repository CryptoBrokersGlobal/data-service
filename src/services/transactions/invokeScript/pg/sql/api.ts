const { compose, defaultTo, merge, pick, pipe, tap } = require('ramda');

const { select, fSelect } = require('./query');
const { pickBindFilters } = require('../../../../../utils/db');

// get — get by id
// mget/search — apply filters
export const createApi = ({ filters: F }: { filters: any }) => ({
  get: (id: string): string => pipe(F.id(id), select, String)(fSelect('desc')),

  mget: (ids: string[]): string =>
    pipe(F.ids(ids), select, String)(fSelect('desc')),

  search: (fValues: Record<string, any> = {}): string => {
    const defaultValues = { limit: 100, sort: 'desc' };

    // no `limit` here, will apply it separately
    const fNames = [
      // tx attributes
      'after',
      'timeStart',
      'timeEnd',
      'sender',
      // specific attributes
      'dapp',
      'function',
      // limit
      'limit',
    ];

    // { [fName]: fValue }
    const withDefaults = compose(pick(fNames), merge(defaultValues))(fValues);
    console.log('fValues ', fValues);
    const fs = pickBindFilters(F, fNames, withDefaults);
    const fQuery = pipe(...fs)(
      fSelect(defaultTo(defaultValues.sort, fValues['sort']))
    );

    return pipe(select, String, tap(console.log))(fQuery);
  },
});
