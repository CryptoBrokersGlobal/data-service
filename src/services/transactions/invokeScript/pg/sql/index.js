const createApi = require('./api');
const filters = require('./filters').default;

module.exports = createApi({ filters });
