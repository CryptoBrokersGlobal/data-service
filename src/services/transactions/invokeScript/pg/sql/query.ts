import * as knex from 'knex';
const pg = knex({ client: 'pg' });

const columnsWithoutFee = [
  // common
  'height',
  'tx_type',
  'id',
  'time_stamp',
  'signature',
  'proofs',
  'tx_version',
  // 'fee',
  'sender',
  'sender_public_key',
  // type-specific
  'dapp',
  'function_name',

  // args
  'arg_type',
  'arg_value_integer',
  'arg_value_boolean',
  'arg_value_binary',
  'arg_value_string',
  'position_in_args',

  // payment
  'amount',
  'asset_id',
  'position_in_payment',
];

export const select = pg({ t: 'txs_16' })
  .select(columnsWithoutFee)
  .select({
    fee: pg.raw('fee * 10^(-8)'),
  })
  .leftJoin({ a: 'txs_16_args' }, 'a.tx_id', 't.id')
  .leftJoin({ p: 'txs_16_payment' }, 'p.tx_id', 't.id');

export const fSelect = pg
  .select('id')
  .min({ time_stamp: 'time_stamp' })
  .from('txs_16')
  .leftJoin({ a: 'txs_16_args' }, 'a.tx_id', 't.id')
  .leftJoin({ p: 'txs_16_payment' }, 'p.tx_id', 't.id')
  .groupBy('id');

export const composeQuery = (filteringQ: knex.QueryBuilder) =>
  select.clone().whereIn('id', pg.select('id').from({ filtered: filteringQ }));
