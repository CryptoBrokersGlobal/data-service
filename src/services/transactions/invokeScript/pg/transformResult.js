const {
  always,
  append,
  assoc,
  assocPath,
  compose,
  cond,
  either,
  evolve,
  groupBy,
  identity,
  isEmpty,
  isNil,
  map,
  merge,
  omit,
  pathOr,
  prop,
  reduce,
  sortBy,
  T,
  uniqWith,
  values,
} = require('ramda');

const getArg = txRow => ({
  type: txRow.arg_type,
  value: txRow[`arg_value_${txRow.arg_type}`],
  positionInArgs: txRow.position_in_args, // for sorting later
});

const getPaymentItem = txRow => ({
  amount: txRow.amount,
  assetId: txRow.asset_id,
  positionInPayment: txRow.position_in_payment, // for sorting later
});

const removeUnnecessaryFromRow = omit([
  'function_name',

  'arg_type',
  'arg_value_integer',
  'arg_value_boolean',
  'arg_value_string',
  'arg_value_binary',
  'position_in_args',

  'amount',
  'asset_id',
  'position_in_payment',
]);

const appendArgsToTx = (tx, row) =>
  compose(
    cond([
      [
        always(!isNil(row.amount)),
        assoc(
          'payment',
          uniqWith(
            (a, b) => a.positionInPayment === b.positionInPayment,
            append(getPaymentItem(row), tx.payment)
          )
        ),
      ],
      [T, identity],
    ]),
    cond([
      [
        always(!isNil(row.arg_type)),
        assocPath(
          ['call', 'args'],
          uniqWith(
            (a, b) => a.positionInArgs === b.positionInArgs,
            append(getArg(row), pathOr([], ['call', 'args'], tx))
          )
        ),
      ],
      [T, identity],
    ]),
    cond([
      [
        always(!isNil(row.function_name)),
        assoc('call', { function: row.function_name, args: [] }),
      ],
      [T, identity],
    ]),
    merge(removeUnnecessaryFromRow(row))
  )(tx);

/**
 * Db returns list of object
 * { ...txAttributes, ...argsAttributes, ...paymentAttributes }
 * Need to restore transaction nested structure, grouping by
 * txAttributes and putting args and payment objects nested into the tx,
 * preserving args and payment order by sorting on `position_in_args` and on `position_in_payment`
 */
const toTxs = cond([
  [either(isNil, isEmpty), always([])],
  [
    T,
    compose(
      // sort by position in tx, then remove it
      map(
        evolve({
          call: evolve({
            args: compose(
              map(omit(['positionInArgs'])),
              sortBy(prop('positionInArgs'))
            ),
          }),
          payment: compose(
            compose(
              map(omit(['positionInPayment'])),
              sortBy(prop('positionInPayment'))
            )
          ),
        })
      ),
      map(reduce(appendArgsToTx, { payment: [] })),
      values,
      groupBy(prop('id'))
    ),
  ],
]);

module.exports = toTxs;
