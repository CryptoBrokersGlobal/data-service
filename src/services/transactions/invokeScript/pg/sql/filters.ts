import * as knex from 'knex';
import * as pointFreeKnex from '../../../../../utils/db/knex';

import * as commonFilters from '../../../_common/sql/filters';

export default {
  ...commonFilters,
  sort: (s: string) => (q: knex.QueryBuilder) => q.clone().orderBy('t.uid', s),
  dapp: pointFreeKnex.where('dapp'),
  function: pointFreeKnex.where('function_name'),
};
