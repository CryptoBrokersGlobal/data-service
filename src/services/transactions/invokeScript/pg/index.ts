import { fromNullable, Maybe } from 'folktale/maybe';
import { head, propEq } from 'ramda';

import { PgDriver } from '../../../../db/driver';
import { matchRequestsResults } from '../../../../utils/db';
import { DbError, toDbError } from '../../../../errorHandling';

import sql from './sql';
import { transformResult } from './transformResult';
import { DbRawInvokeScriptTx, RawInvokeScriptTx } from '../types';

export default {
  get: (pg: PgDriver) => (id: string) =>
    pg
      .any<DbRawInvokeScriptTx>(sql.get(id))
      .map<RawInvokeScriptTx[]>(transformResult)
      .map<RawInvokeScriptTx>(head)
      .map<Maybe<RawInvokeScriptTx>>(fromNullable)
      .mapRejected<DbError>(e =>
        toDbError(
          { request: 'transactions.invokeScript.get', params: id },
          e.error
        )
      ),

  mget: (pg: PgDriver) => (ids: string[]) =>
    pg
      .any<DbRawInvokeScriptTx>(sql.mget(ids))
      .map<RawInvokeScriptTx[]>(transformResult)
      .map<Maybe<RawInvokeScriptTx>[]>(matchRequestsResults(propEq('id'), ids))
      .mapRejected<DbError>(e =>
        toDbError(
          { request: 'transactions.invokeScript.mget', params: ids },
          e.error
        )
      ),

  search: (pg: PgDriver) => (filters: Record<string, any>) =>
    pg
      .any<DbRawInvokeScriptTx>(sql.search(filters))
      .map<RawInvokeScriptTx[]>(transformResult)
      .mapRejected<DbError>(e =>
        toDbError(
          {
            request: 'transactions.invokeScript.search',
            params: filters,
            message: e.error.message,
          },
          e.error
        )
      ),
};
