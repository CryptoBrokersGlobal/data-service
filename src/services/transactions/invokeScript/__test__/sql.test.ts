import { search } from '../pg/sql';
const filterValues = [
  {
    dapp: '3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve',
  },
  {
    function: 'deposit',
  },
  {
    dapp: 'b3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve',
    function: 'deposit',
  },
];

describe('Sql search by type-specific filters', () => {
  it('supports dapp filter', () => {
    expect(search(filterValues[0])).toMatchSnapshot();
  });

  it('supports function filter', () => {
    expect(search(filterValues[1])).toMatchSnapshot();
  });

  it('supports dapp and function filters', () => {
    expect(search(filterValues[2])).toMatchSnapshot();
  });
});
