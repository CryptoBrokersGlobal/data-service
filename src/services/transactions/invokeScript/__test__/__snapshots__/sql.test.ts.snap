// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sql search by type-specific filters supports dapp and function filters 1`] = `"select \\"height\\", \\"tx_type\\", \\"id\\", \\"time_stamp\\", \\"signature\\", \\"proofs\\", \\"tx_version\\", \\"sender\\", \\"sender_public_key\\", \\"dapp\\", \\"function_name\\", \\"arg_type\\", \\"arg_value_integer\\", \\"arg_value_boolean\\", \\"arg_value_binary\\", \\"arg_value_string\\", \\"position_in_args\\", \\"amount\\", \\"asset_id\\", \\"position_in_payment\\", fee * 10^(-8) as \\"fee\\" from \\"txs_16\\" as \\"t\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\" where \\"id\\" in (select \\"id\\" from (select \\"id\\", min(\\"time_stamp\\") as \\"time_stamp\\" from \\"txs_16\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\" where \\"dapp\\" = 'b3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve' and \\"function_name\\" = 'deposit' group by \\"id\\" order by \\"time_stamp\\" desc, \\"id\\" desc limit 100) as \\"filtered\\") order by \\"time_stamp\\" desc, \\"id\\" desc"`;

exports[`Sql search by type-specific filters supports dapp filter 1`] = `"select \\"height\\", \\"tx_type\\", \\"id\\", \\"time_stamp\\", \\"signature\\", \\"proofs\\", \\"tx_version\\", \\"sender\\", \\"sender_public_key\\", \\"dapp\\", \\"function_name\\", \\"arg_type\\", \\"arg_value_integer\\", \\"arg_value_boolean\\", \\"arg_value_binary\\", \\"arg_value_string\\", \\"position_in_args\\", \\"amount\\", \\"asset_id\\", \\"position_in_payment\\", fee * 10^(-8) as \\"fee\\" from \\"txs_16\\" as \\"t\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\" where \\"id\\" in (select \\"id\\" from (select \\"id\\", min(\\"time_stamp\\") as \\"time_stamp\\" from \\"txs_16\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\" where \\"dapp\\" = '3N92gNUHh6jnZmZtdgECTm3cNAZb1Zei6Ve' group by \\"id\\" order by \\"time_stamp\\" desc, \\"id\\" desc limit 100) as \\"filtered\\") order by \\"time_stamp\\" desc, \\"id\\" desc"`;

exports[`Sql search by type-specific filters supports function filter 1`] = `"select \\"height\\", \\"tx_type\\", \\"id\\", \\"time_stamp\\", \\"signature\\", \\"proofs\\", \\"tx_version\\", \\"sender\\", \\"sender_public_key\\", \\"dapp\\", \\"function_name\\", \\"arg_type\\", \\"arg_value_integer\\", \\"arg_value_boolean\\", \\"arg_value_binary\\", \\"arg_value_string\\", \\"position_in_args\\", \\"amount\\", \\"asset_id\\", \\"position_in_payment\\", fee * 10^(-8) as \\"fee\\" from \\"txs_16\\" as \\"t\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\" where \\"id\\" in (select \\"id\\" from (select \\"id\\", min(\\"time_stamp\\") as \\"time_stamp\\" from \\"txs_16\\" left join \\"txs_16_args\\" as \\"a\\" on \\"a\\".\\"tx_id\\" = \\"t\\".\\"id\\" left join \\"txs_16_payment\\" as \\"p\\" on \\"p\\".\\"tx_id\\" = \\"t\\".\\"id\\" where \\"function_name\\" = 'deposit' group by \\"id\\" order by \\"time_stamp\\" desc, \\"id\\" desc limit 100) as \\"filtered\\") order by \\"time_stamp\\" desc, \\"id\\" desc"`;
